docs/user/search.rst

```diff
@@ -0,0 +1,50 @@
+.. _search:
+
+Using the search feature
+========================
+
+Pytube includes functionality to search YouTube and return results almost
+identical to those you would find using the search bar on YouTube's website.
+The integration into pytube means that we can directly provide you with
+YouTube objects that can be inspected and dowloaded, instead of needing to do
+additional processing.
+
+Using the Search object is really easy::
+
+    >>> from pytube import Search
+    >>> s = Search('YouTube Rewind')
+    >>> len(s.results)
+    17
+    >>> s.results
+    [\
+        <pytube.__main__.YouTube object: videoId=YbJOTdZBX1g>, \
+        <pytube.__main__.YouTube object: videoId=PKtnafFtfEo>, \
+        ...\
+    ]
+    >>> 
+
+Due to the potential for an endless stream of results, and in order to prevent
+a user from accidentally entering an infinite loop of requesting additional
+results, the ``.results`` attribute will only ever request the first set of
+search results. Additional results can be explicitly requested by using the
+``.get_next_results()`` method, which will append any additional results to
+the ``.results`` attribute::
+
+    >>> s.get_next_results()
```

Comment:
would suggest returning an iterator which has a buffer of $k

https://wiki.python.org/moin/Iterator

Reply:
I like iterators, but I don't think it makes sense in the case of search queries. With e.g. playlists, channel videos, etc., there's a definitive end, so iterators make sense. I don't know for certain if search queries have an end, because youtube likes to be like "hey, maybe you'll want this as well?", and incorporates that into the endpoint.

You could make an argument for using a generator here, since those are designed to support a system for continuously generating output. In fact, I've used a wrapper around generators elsewhere in the codebase to improve our performance. However, I think we should follow more closely with the innertube API's intended use case -- you request a few at first, and if you need more, then you ask for more. Additionally, using generators isn't a great user experience in python. We could do it, but writing something like
```python
s = Search(query)
count = 0
while count < 20:
    curr_vid = next(s.videos)
```
just doesn't really align with how the rest of the library works.

Reply:
@tfdahlin  would +1 to generator.

> However, I think we should follow more closely with the innertube API's intended use case -- you request a few at first, and if you need more, then you ask for more. Additionally, using generators isn't a great user experience in python. We could do it, but writing something like

Sorry don't think i'm understanding - we can mask this using a buffer in the generator implementation 
```
buffer = deque()
def __next__( ...):
   if buffer:
       return buffer.pop() 
    if getMore := innerTube.moreResults():
          buffer = getMore[1:]
          return getMore[0]
     return StopIteration
```